thread-manager/src/tokio_runtime.rs

```diff
@@ -0,0 +1,185 @@
+            namespace: format!("thread-manager-tokio-{}", &base_name).leak(), // no workaround, metrics crate will only consume 'static str
+            total_threads_cnt: cfg.worker_threads as u64,
+            active_threads_cnt: AtomicU64::new(
+                (num_workers.wrapping_add(cfg.max_blocking_threads)) as u64,
+            ),
+        });
+        builder
+            .event_interval(cfg.event_interval)
+            .thread_name_fn(move || {
+                let id = atomic_id.fetch_add(1, Ordering::Relaxed);
+                format!("{}-{}", base_name, id)
+            })
+            .on_thread_park({
+                let counters = counters.clone();
+                move || {
+                    counters.on_park();
+                }
+            })
+            .on_thread_unpark({
+                let counters = counters.clone();
+                move || {
+                    counters.on_unpark();
+                }
+            })
+            .thread_stack_size(cfg.stack_size_bytes)
+            .enable_all()
+            .max_blocking_threads(cfg.max_blocking_threads);
+
+        //keep borrow checker happy and move these things into the closure
+        let c = cfg.clone();
+        let chosen_cores_mask = Mutex::new(chosen_cores_mask);
+        builder.on_thread_start(move || {
+            let cur_thread = std::thread::current();
+            let _tid = cur_thread
+                .get_native_id()
+                .expect("Can not get thread id for newly created thread");
+
+            apply_policy(
+                &c.core_allocation,
+                parse_policy(&c.policy),
+                c.priority,
+                &chosen_cores_mask,
+            );
+        });
+        Ok(TokioRuntime {
+            tokio: builder.build()?,
+            config: cfg.clone(),
+            counters,
+        })
+    }
+
+    /// Makes test runtime with 2 threads, only for unittests
+    #[cfg(feature = "dev-context-only-utils")]
+    pub fn new_for_tests() -> Self {
+        let cfg = TokioConfig {
+            worker_threads: 2,
+            ..Default::default()
+        };
+        TokioRuntime::new("solNetTest".to_owned(), cfg.clone())
+            .expect("Failed to create Tokio runtime for tests")
+    }
+}
+
+///Internal counters to keep track of worker pool utilization
+#[derive(Debug)]
+pub struct ThreadCounters {
+    pub namespace: &'static str,
+    pub total_threads_cnt: u64,
+    pub active_threads_cnt: AtomicU64,
+}
+
+impl ThreadCounters {
+    pub fn on_park(&self) {
+        self.active_threads_cnt.fetch_sub(1, Ordering::Relaxed);
+    }
+
+    pub fn on_unpark(&self) {
+        self.active_threads_cnt.fetch_add(1, Ordering::Relaxed);
+    }
+}
+
+async fn metrics_sampler(counters: Arc<ThreadCounters>, period: Duration) {
+    let mut interval = tokio::time::interval(period);
+    loop {
+        interval.tick().await;
+        let active = counters.active_threads_cnt.load(Ordering::Relaxed) as i64;
```

Comment:
should we use `SeqCst` here?

Reply:
there is no need, we are not using this value to compute any memory addresses. Even if this is slightly "out of date" this is unlikely to matter as the value is only for metrics report.